### DISCUSSION TITLE
How to get the current time on CAN Tx to calculate Rx response timing

### DISCUSSION THREAD

**Discussion by sveinse** [hardbyte/python-can#1863] [Q&A]
I'm interested in capturing the response time on CAN bus requests. The Rx frames have `timestamp` attributes, but the Tx frames my code generates does not have a timestamp. Is there a way to set the current time when the Tx message was sent? Alternatively, is there a get_current_time() function that can be called to fetch the time (which is using the same epoch and time mechanisms as the receivers)?

I'm on Windows using PCAN.

**Comment by zariiii9003** [comment]
Did you instantiate the `PcanBus` with `receive_own_messages=True`? I'm not sure how the PCAN API handles this, but i think the `recv()` method should return your Tx messages with a timestamp.

**Comment by sveinse** [comment]
I'm not getting the responses back with `receive_own_messages=True` with the PCAN I have. I looked into the pcan code, and it seems the setting have a operation against the PCAN driver, but it evidently doesn't do anything.

**Reply by zariiii9003** [reply]
According to [PCAN docs](https://www.peak-system.com/produktcd/Develop/PC%20interfaces/Windows/PCAN-Basic%20API/PCAN-Parameter_Documentation.pdf), setting `PCAN_ALLOW_ECHO_FRAMES` to `PCAN_PARAMETER_ON` should be sufficient.

Maybe it's not supported by your device or driver version?

**Comment by sveinse** [comment]
I tried using `datetime.now().timestamp()` to set my own timestamp on Tx, but I get 1.3 second offset from the true values as measured by HW trace. Without a way to set `.timestamp` on Tx messages or to read the current timestamp from the PCAN driver, I don't think I'm able to use the `.timestamp` field in the Rx messages.

**Comment by sveinse** [comment]
It turns out had a simple typo throughout the code: `receive_own_message = True`. Unfortunately python-can doesn't disagree with using that, except it doesn't work obviously. So I ended up with receiving my own messages to retrieve its timestamp which works perfectly. Closing the discussion.
